simulation/halsim_ws_core/src/main/native/cpp/WSProvider_SimDevice.cpp

```diff
@@ -0,0 +1,209 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2017-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "WSProvider_SimDevice.h"
+
+#include <hal/Ports.h>
+
+namespace wpilibws {
+
+HALSimWSProviderSimDevice::~HALSimWSProviderSimDevice() { CancelCallbacks(); }
+
+void HALSimWSProviderSimDevice::OnNetworkConnected(
+    std::shared_ptr<HALSimBaseWebSocketConnection> ws) {
+  auto storedWS = m_ws.lock();
+
+  if (ws == storedWS) {
+    return;
+  }
+
+  // Otherwise, run the disconnection code first so we get
+  // back to a clean slate (only if the stored websocket is
+  // not null)
+  if (storedWS) {
+    OnNetworkDisconnected();
+  }
+
+  m_ws = ws;
+
+  m_simValueCreatedCbKey = HALSIM_RegisterSimValueCreatedCallback(
+      m_handle, this, HALSimWSProviderSimDevice::OnValueCreatedStatic, 1);
+
+  {
+    auto it = m_valueHandles.begin();
+    auto end = m_valueHandles.end();
+
+    std::shared_lock lock(m_vhLock);
+    HAL_Value value;
+    for (; it != end; ++it) {
+      HAL_GetSimValue(it->second->handle, &value);
+      OnValueChanged(it->second.get(), &value);
```

Comment:
Lines 36-46 can be deleted, since the ValueCreatedCallback will do all of this work automatically for you?

Reply:
Good catch! This is why I need to drink more coffee before writing code...